File: backend/app/schemas/seasonality.py
"""Seasonality API schemas — monthly-return distribution (read-only)."""

from __future__ import annotations

from datetime import date

from pydantic import BaseModel


class MonthSeasonalityOut(BaseModel):
    month: int                       # 1..12
    n: int                           # yearly observations for this month
    positive: int
    pct_positive: float | None       # None when n == 0
    avg_return_pct: float | None
    median_return_pct: float | None
    best_return_pct: float | None
    worst_return_pct: float | None
    avg_positive_pct: float | None
    avg_negative_pct: float | None


class SeasonalityOut(BaseModel):
    stock_id: int
    total_observations: int          # total monthly returns counted
    years_covered: int               # distinct years contributing a return
    first_month: date | None
    last_month: date | None
    months: list[MonthSeasonalityOut]  # always 12, ordered month 1..12

File: frontend/src/lib/api/seasonality.ts
import { api } from './client'

export interface MonthSeasonality {
  month: number // 1..12
  n: number // yearly observations for this month
  positive: number
  pct_positive: number | null // null when n === 0
  avg_return_pct: number | null
  median_return_pct: number | null
  best_return_pct: number | null
  worst_return_pct: number | null
  avg_positive_pct: number | null
  avg_negative_pct: number | null
}

export interface Seasonality {
  stock_id: number
  total_observations: number
  years_covered: number
  first_month: string | null // ISO date (first of month)
  last_month: string | null
  months: MonthSeasonality[] // always 12, ordered month 1..12
}

export const seasonalityApi = {
  get: (stockId: number, token: string) =>
    api.get<Seasonality>(`/seasonality/${stockId}`, token),
}
